src/charm.py

```diff
@@ -0,0 +1,392 @@
+        if result := re.search(r"[Vv]ersion:?\s*(\S+)", version_output):
+            return result.group(1)
+        return None
+
+    @property
+    def tempo_endpoint(self) -> Optional[str]:
+        """Tempo endpoint for charm tracing."""
+        if self.tracing.is_ready():
+            return self.tracing.otlp_http_endpoint()
+        else:
+            return None
+
+    @property
+    def server_cert_path(self) -> Optional[str]:
+        """Server certificate path for tls tracing."""
+        return LOKI_CERT_FILE
+
+    # === UTILITY METHODS === #
+
+    def _update_loki_cluster(self):
+        """Share via loki-cluster all information we need to publish."""
+        self.loki_cluster.publish_unit_address(socket.getfqdn())
+        if self.unit.is_leader() and self._loki_roles:
+            logger.info(f"publishing roles: {self._loki_roles}")
+            print("*" * 100)
+            print(self._loki_roles)
+            print("*" * 100)
+            self.loki_cluster.publish_app_roles(self._loki_roles)
+
+    def _update_config(self):
+        """Update the loki config and restart the workload if necessary."""
+        restart = any(
+            [
+                self._update_tls_certificates(),
+                self._update_loki_config(),
+                self._set_pebble_layer(),
+            ]
+        )
+
+        if restart:
+            self.restart()
+
+    def _set_pebble_layer(self) -> bool:
+        """Set Pebble layer.
+
+        Returns: True if Pebble layer was added, otherwise False.
+        """
+        if not self._container.can_connect():
+            return False
+        if not self._loki_roles:
+            return False
+
+        current_layer = self._container.get_plan()
+        new_layer = self._pebble_layer
+
+        if (
+            "services" not in current_layer.to_dict()
+            or current_layer.services != new_layer["services"]
+        ):
+            self._container.add_layer(self._name, new_layer, combine=True)  # pyright: ignore
+            return True
+
+        return False
+
+    def _set_ports(self):
+        """Open necessary (and close no longer needed) workload ports."""
+        planned_ports = {
+            OpenedPort("tcp", self._port),
+        }
+        actual_ports = self.unit.opened_ports()
+
+        # Ports may change across an upgrade, so need to sync
+        ports_to_close = actual_ports.difference(planned_ports)
+        for p in ports_to_close:
+            self.unit.close_port(p.protocol, p.port)
+
+        new_ports_to_open = planned_ports.difference(actual_ports)
+        for p in new_ports_to_open:
+            self.unit.open_port(p.protocol, p.port)
+
+    def _update_tls_certificates(self) -> bool:
+        if not self._container.can_connect():
+            return False
+
+        ca_cert_path = Path("/usr/local/share/ca-certificates/ca.crt")
+
+        if cert_secrets := self.loki_cluster.get_cert_secret_ids():
```

Comment:
Can we obtain a secret by label without going through id?

Reply:
Yes we can.
I would refactor this once we have coordinator and worker in a deployable state.
Let's follow this in https://github.com/canonical/loki-worker-k8s-operator/issues/3